text/0219-ecs-service-extensions.md

```diff
@@ -0,0 +1,503 @@
+like Virtual Node and Virtual Service. There is no single place for all of this cross
+construct logic to live in the current construct architecture.
+
+Last but not least the existing approach to L3 constructs has an `Ec2` and a `Fargate`
+version of each L3 construct. Ideally there would be a way for customers to create a
+service which is agnostic to whether it is deployed on EC2 or Fargate.
+
+## Tenants
+
+To solve the problems with ECS construct implemention we propose adding a new `Service`
+class with the following goals:
+
+- __One service class to learn__: Instead of a collection of many different L3 classes
+  that all have different names there is a single class which can implement any number
+  of different forms depending on what the customer wants. Customers don't need to learn
+  the names for different L3 constructs. All ECS architectures start the same way: with
+  the same `Service` class.
+- __Easy to extend and customize__: It is easy to extend the `Service` class with new
+  behaviors and features, in any combination required. The `Service` class comes with
+  batteries included: a set of easy to use extensions created and maintained by the AWS team.
+  However there are also clear instructions and examples of how customers can create their
+  own custom extensions. Third parties will be enabled to also create and distribute
+  extensions for their own services, for example a provider like Datadog could easily
+  build an extension that automatically adds Datadog agent to an ECS service.
+- __Extensions self configure and self provision__: Service extensions provision their
+  own resources as well as configuring the ECS service to use those resources. For example
+  a load balancer extension creates its own load balancer and also attaches the ECS service
+  to the load balancer. An App Mesh extension creates its own App Mesh virtual service and
+  virtual node as well as configuring the ECS service and ECS task definition to have the
+  right Envoy sidecar and proxy settings.
+- __Extensions are aware of each other__: There are no hidden "gotchas" when enabling a
+  service extension. Extensions know of each other's existence. For example the application's
+  container is aware that there is an Envoy proxy and that it must wait until the Envoy
+  proxy has started before the app starts.
+- __Service class is agnostic to EC2 vs Fargate__: Rather than choosing the capacity strategy
+  in the class name customers just create a service, and add it to an environment. The
+  service automatically modifies its settings as needed to adjust to the capacity available
+  in the environment.
+
+# Design Summary
+
+In this proposal we introduce two new constructs:
+
+- `Environment` - A simple construct which wraps up a VPC, cluster, and other top level
+  resources needed for ECS services to run on an AWS account
+- `Service` - A single application that runs inside an `Environment`
```

Comment:
- The term "Environment" in the CDK is currently used to represent an AWS account+region. Can we come up with a term that does not overload?
- What's the relation between Service and AWS::ECS::Service? Are we overloading the term or there a 1:1 mapping between this L3 Service and the underlying AWS::ECS::Service.

Reply:
I agree this may be a bit confusing, however I would argue that "environment" is a sufficiently generic term which is used in many places, by many AWS services, so there should already be a user expectation of "which environment are we talking about". In this case this `Environment` and `Service` is deliberately designed to map to the "environment" and "service" in the AWS Copilot project. One of the goals is for this construct to feel familiar, approachable, and consistent to consumers of the new higher level tooling we are building for ECS (not just in CDK, but our CLI, console revamp and other upcoming projects)

There is a 1:1 mapping between this `Service` construct an an underlying `AWS::ECS::Service` however this higher level L3 Service also wraps up and abstracts away the `AWS::ECS::TaskDefinition` and other required resources for a service to run and deploy